spring-cloud-gcp-data-firestore/src/main/java/org/springframework/cloud/gcp/data/firestore/util/ObservableReactiveUtil.java

```diff
@@ -68,6 +72,21 @@ private ObservableReactiveUtil() {
 		});
 	}
 
+	public static <ResponseT, RequestT, A> Flux<ResponseT> streamingBidirectionalCall(
+			Function<StreamObserver<ResponseT>, StreamObserver<RequestT>> remoteCall,
+			Flux<A> inputs, BiFunction<A, Flux<ResponseT>, Mono<RequestT>> requestFunc) {
+
+		AtomicReference<StreamObserver<RequestT>> requestObserver = new AtomicReference<>();
+		Flux<ResponseT> responses = (Flux<ResponseT>) streamingCall(
+				obs -> requestObserver.set(remoteCall.apply((StreamObserver<ResponseT>) obs))).cache();
+
+		return inputs
+				.flatMap((A input) -> requestFunc.apply(input, responses)
+						.doOnNext(request -> requestObserver.get().onNext(request)))
+				.doOnComplete(() -> requestObserver.get().onCompleted())
+				.thenMany(responses);
```

Comment:
I think this causes the two direction not to be asynchronous.
If I'm reading this correctly, you will not emit any responses until the requests stream is completed.

Reply:
This is actually exactly the point.

Otherwise you will return before the final write has taken effect.

You still gain benefits because you are never blocked .the waiting is handled by the stream's own thread.

You absolutely must wait until the final write confirmation if you want strong consistency. Otherwise you have dirty eventual consistency like data store.

Reply:
Of course I could be mistaken :D

Reply:
Why couldn't you wait for the response flux to complete to make sure that all writes happened? The point is that you should be able to be notified that the first write has completed before the last write has completed or even published. That would be asynchronous.

Reply:
https://github.com/spring-cloud/spring-cloud-gcp/issues/1813